lib/persisted-model.js

```diff
@@ -651,6 +673,22 @@ module.exports = function(registry) {
 
     setRemoting(PersistedModel, 'replaceOrCreate', replaceOrCreateOptions);
 
+    setRemoting(PersistedModel, 'upsertWithWhere', {
+      aliases: ['patchOrCreateWithWhere'],
+      description: 'Update an existing model instance or ' +
+      'insert a new one into the data source based on the ' +
+      'where criteria.',
+      accessType: 'WRITE',
+      accepts: [
+        { arg: 'where', type: 'object', http: { source: 'query' },
+          description: 'Criteria to match model instances' },
+        { arg: 'data', type: 'object', http: { source: 'body' },
+          description: 'An object of model property name/value pairs' },
+      ],
+      returns: { arg: 'data', type: typeName, root: true },
+      http: { verb: 'put', path: '/upsertWithWhere' },
```

Comment:
I think we usually use `POST` for non-root endpoints like `/upsertWithWhere`.

Alternatively, we can use `PATCH` to be consistent with `upsert` that's mapped to `PATCH /` by default since LB 3.0.